url3426a97ee4_rtreffer_minidns_tgz-pJ8-de_measite_minidns_minidnsrepl_MiniDnsReplJ8/src/de/measite/minidns/dnssec/DNSSECClient.java

```diff
@@ -0,0 +1,555 @@
+        DNSMessage dnsMessage = super.query(q);
+        DNSSECMessage dnssecMessage = performVerification(q, dnsMessage);
+        return dnssecMessage;
+    }
+
+    public DNSSECMessage queryDnssec(Question q) throws IOException {
+        DNSMessage dnsMessage = super.query(q);
+        DNSSECMessage dnssecMessage = performVerification(q, dnsMessage);
+        return dnssecMessage;
+    }
+
+    private DNSSECMessage performVerification(Question q, DNSMessage dnsMessage) throws IOException {
+        if (dnsMessage == null) return null;
+
+        // At this state, a DNSMessage is never authentic!
+        if (dnsMessage.authenticData) {
+            dnsMessage = dnsMessage.asBuilder().setAuthenticData(false).build();
+        }
+
+        Set<UnverifiedReason> unverifiedReasons = verify(dnsMessage);
+
+        return createDnssecMessage(dnsMessage, unverifiedReasons);
+    }
+
+    private DNSSECMessage createDnssecMessage(DNSMessage dnsMessage, Set<UnverifiedReason> result) {
+        List<Record<? extends Data>> answers = dnsMessage.answerSection;
+        List<Record<? extends Data>> nameserverRecords = dnsMessage.authoritySection;
+        List<Record<? extends Data>> additionalResourceRecords = dnsMessage.additionalSection;
+        Set<Record<RRSIG>> signatures = new HashSet<>();
+        Record.filter(signatures, RRSIG.class, answers);
+        Record.filter(signatures, RRSIG.class, nameserverRecords);
+        Record.filter(signatures, RRSIG.class, additionalResourceRecords);
+        DNSMessage.Builder messageBuilder = dnsMessage.asBuilder();
+        if (stripSignatureRecords) {
+            messageBuilder.setAnswers(stripSignatureRecords(answers));
+            messageBuilder.setNameserverRecords(stripSignatureRecords(nameserverRecords));
+            messageBuilder.setAdditionalResourceRecords(stripSignatureRecords(additionalResourceRecords));
+        }
+        return new DNSSECMessage(messageBuilder, signatures, result);
+    }
+
+    private static List<Record<? extends Data>> stripSignatureRecords(List<Record<? extends Data>> records) {
+        if (records.isEmpty()) return records;
+        List<Record<? extends Data>> recordList = new ArrayList<>(records.size());
+        for (Record<? extends Data> record : records) {
+            if (record.type != TYPE.RRSIG) {
+                recordList.add(record);
+            }
+        }
+        return recordList;
+    }
+
+    @Override
+    protected boolean isResponseCacheable(Question q, DNSMessage dnsMessage) {
+        return super.isResponseCacheable(q, dnsMessage);
+    }
+
+    private Set<UnverifiedReason> verify(DNSMessage dnsMessage) throws IOException {
+        if (!dnsMessage.answerSection.isEmpty()) {
+            return verifyAnswer(dnsMessage);
+        } else {
+            return verifyNsec(dnsMessage);
+        }
+    }
+
+    private Set<UnverifiedReason> verifyAnswer(DNSMessage dnsMessage) throws IOException {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 12. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 163-186 into a separate method.